llvm/lib/CodeGen/BranchFolding.cpp

```diff
@@ -1035,6 +1040,9 @@ bool BranchFolder::TailMergeBlocks(MachineFunction &MF) {
   if (MergePotentials.size() >= 2)
     MadeChange |= TryTailMergeBlocks(nullptr, nullptr, MinCommonTailLength);
 
+  if (TailMergeOnlyBBsWithoutSucc)
```

Comment:
Hello !

> this option can reduce code size in case function has multiple return points that can be merged.

Is the driver that sets the option when optimizing for size?

Could we instead check if _optimizing for code size_ and _the function has multiple return points that can be merged_ instead of using an option?

I'd also expect that if a branch edge [is hot](https://llvm.org/doxygen/classllvm_1_1MachineBranchProbabilityInfo.html#a6cd25f432a77253606cc6387d0057833) this transformation would still be applied (even if optimizing for size).

Reply:
Sorry for the confusion, but this option can reduce code size when comparing application size when `-enable-tail-merge=false` is passed, not in usual cases.
I'm working on a downstream target where jumps are not cheap and BranchFolding can create a lot of them in hot paths. For applications where performance is pivotal, I found that disabling tail merge (`-enable-tail-merge=false`) is boosting performance a lot, and saw that we can reduce code size if tail merge was performed for returning BBs, which in our case is a cold path. Because of this, I introduced this option to cover that case.

Reply:
Ok, so if I understand this would optimize for performance in your target and not for code-size.

In that case, have you thought about making the `enable-tail-merge` (and the value in the constructor of `BranchFolder`) into an `enum` covering your case ?

I'd not be against a target hook returning `true` if "unconditional-jumps-are-expensive" that helps control this behavior.

Reply:
I went with a 2nd approach to create target hook, so we can use it in a downstream target. PTAL.